runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,219 @@
+-- Create a file watcher, each watcher is identified by the name of the file that it
+-- watches.
+
+local uv = vim.loop
+local Watcher = {}
+Watcher.__index = Watcher
+local WatcherList = {}
+local check_handle = nil
+local in_focus = 1
+
+-- Checks if a buffer should have a watcher attached to it.
+local function valid_buf(bufnr)
+  if not vim.api.nvim_buf_is_valid(bufnr) then
+    return false
+  end
+
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  return buflisted or buftype == '' or buftype == 'acwrite'
+end
+
+-- Callback for the check handle, checks if there are pending notifications
+-- for any watcher, and handles them as per the value of the `fcnotify`
+-- option.
+local function check_notifications()
+  for _, watcher in pairs(WatcherList) do
+    if watcher.pending_notifs and valid_buf(watcher.bufnr) then
+      vim.api.nvim_command(string.format('checktime %d', watcher.bufnr))
+      watcher.pending_notifs = false
+    end
+  end
+end
+
+function Watcher.set_focus(focus)
+  in_focus = focus
+  if in_focus then
+   check_notifications()
+  end
+end
+
+local function fs_event_start(bufnr)
+  local watcher = WatcherList[bufnr]
+  watcher.handle = uv.new_fs_event()
+  watcher.handle:start(watcher.fpath, {}, vim.schedule_wrap(function(...)
+    watcher:on_change(...)
+  end))
+end
+
+local function set_mechanism(option_type, bufnr)
+  if option_type == 'global' then
+    local option = vim.api.nvim_get_option('filechangenotify')
+    if vim.tbl_contains(vim.split(option, ','), 'watcher') then
+      for _, watcher in pairs(WatcherList) do
+        watcher._start_handle = fs_event_start
+      end
+    else
+      for _, watcher in pairs(WatcherList) do
+        watcher._start_handle = nil
+      end
+    end
+  elseif option_type == 'local' then
+    bufnr = bufnr or vim.api.nvim_get_current_buf()
+    local watcher = WatcherList[bufnr]
+    local status, option = pcall(vim.api.nvim_buf_get_option, bufnr, 'filechangenotify')
+    if not status then
+      option = vim.api.nvim_get_option('filechangenotify')
+    end
+    if vim.tbl_contains(vim.split(option, ','), 'watcher') then
+      watcher._start_handle = fs_event_start
+    else
+      watcher._start_handle = nil
+    end
+  end
+end
+
+--- Creates and initializes a new watcher object with the given filename.
+---
+--@param bufnr: (required, number) The buffer number of the buffer that
+--- is to be watched.
+function Watcher:new(bufnr)
+  vim.validate{bufnr = {bufnr, 'number', false}}
+  -- get full path name for the file
+  local fname = vim.api.nvim_buf_get_name(bufnr)
+  local fpath = vim.fn.fnamemodify(fname, ':p')
+  local w = {bufnr = bufnr, fname = fname, fpath = fpath,
+       handle = nil, pending_notifs = false}
+  w._start_handle = nil
```

Comment:
can just put `_start_handle = nil` in the line above with the other declarations.

I also prefer only one declaration per line, it is a lot easier to read that way.